### DISCUSSION TITLE
[BUG]: The progress bar stops at a certain percentage and warm_start doesn't seem to be working

### DISCUSSION THREAD

**Discussion by Angelld23** [MilesCranmer/PySR#956] [Q&A]
### What happened?

Hello

Every time I start a new run, the model start having progress for a couple of minutes, but after 30 minutes or around 11% of progress it stops improving. I tried leaving it there for a couple hours to see if it just that the search time increased exponentially, but it doesn't look like. I would like to know if this might be a bug or maybe is a problem of my laptop. Also, When I try to run again starting from the last checkpoint (warm_start=True) it doesn't look like it works because it starts from a new checkpoint with different equations with higher loss.

System: Hp Laptop

Edition: Windows 11 Enterprise
Version: 24H2
Processor: 13th Gen Intel(R) Core(TM) i7-1360P   2.20 GHz
RAM: 32.0 GB (31.6 GB usable)
System Type: 64-bit operating system, x64-based processor

Python: 3.11.9
PySR version: 1.5.8

Code running on Jupyter Kernel (notebook) in VS-Code

### Version

1.5.8

### Operating System

Windows

### Package Manager

pip

### Interface

Jupyter Notebook

### Relevant log output

```shell

```

### Extra Info

```python
loss_function = """
function f(tree, dataset::Dataset{T,L}, options) where {T,L}
    ypred, completed = eval_tree_array(tree, dataset.X, options)
    if !completed
        return L(Inf)
    end
    y = dataset.y
    return sqrt(sum(i -> (ypred[i] - y[i])^2, eachindex(y)))
end
"""

model = PySRRegressor(
    niterations=200,  # < Increase me for better results
    ncycles_per_iteration = 380,
    populations=250,
    population_size=250,
    binary_operators=["*", "/", "-", "+"],
    unary_operators=[
        "exp",
        "sqrt",
        "square",
        "log10",
        "cube",
    ],
    # ^ Define operator for SymPy as well
    warm_start=True,
    maxsize=40,
    loss_function=loss_function,
    adaptive_parsimony_scaling =1000,
    fraction_replaced_hof =0.1,
    select_k_features = 5,
    parsimony = 0.0001,
    weight_optimize = 0.001,
    nested_constraints = {
        "exp": {"exp": 0},
        "sqrt": {"sqrt": 0},
        "square": {"square": 0},
        "log10": {"log10": 0},
        "cube": {"cube": 0},
    },
    output_directory=r'C:\Users\xxxxxx',
)
model.fit(X_train, y_train)
 ```

**Comment by MilesCranmer** [comment]
Just to check: does it continue the search (just no longer finding better results), or does it completely freeze at 11%?

Also please use triple backticks to paste code in comments as it’s easier to read that way

**Reply by Angelld23** [reply]
It completely freezes in the last update it gave me at around less than 11%. The output in the jupyter notebook doesn't get updated with the new iterations and neither the checkpoint files. They all stay the same, no matter how many hours or how many times I try to re-run the code. 

I will use triple backticks next time.

**Reply by MilesCranmer** [reply]
1. Does the task manager show Julia is still running? Or did it quit? (Is it using the CPU at all if it's still active?)
2. If you quit (ctrl-c) or kill the kernel, can you paste the stacktrace/error message/logs? It should say where it is stuck.

**Reply by Angelld23** [reply]
Whenever I run the jupyter notebook, I only see that the vs-code .ipynb file is running in task manager. Even from the start I never see Julia showing up in task manager. Still I noticed that the ipynb file is using almost to no CPU so the file is probably not running although in the jupyter notebook it shows like it still running because I can see the running cell loading and the minutes/seconds elapsed still counting.

I am unable to quit or kill the kernel because it always ask me to restart it every time I interrup the run. I also try to trace every log from the jupyter kernel and didn't get much information on what was happening. The notebook cell shows up as if it is still running but no progress or updates on the checkpoints is sent. It's being 3 hours and theres has been no progress. I will try to use it on python file .py to see if it still the same. Below is the kernel output log:

```python
10:22:36.634 [trace] Got Kernel info
10:22:36.634 [trace] End running kernel initialization, now waiting for idle
10:22:36.634 [trace] Waiting for idle on (kernel): 8f8f237e-0e26-4d8c-8d4d-a37209cce833 -> idle
10:22:36.635 [trace] Finished waiting for idle on (kernel): 8f8f237e-0e26-4d8c-8d4d-a37209cce833 -> idle
10:22:36.635 [trace] End running kernel initialization, session is idle
10:22:36.635 [trace] Registering commtarget jupyter.widget
10:25:02.221 [debug] Jupyter completion time: 18
10:25:02.223 [debug] Execution of code ms-toolsai.jupyter-2 completed in 19ms
10:25:03.747 [debug] Execution of code ms-toolsai.jupyter-3 completed in 18ms
10:25:03.747 [debug] Jupyter completion time: 18
10:25:03.863 [debug] Execution of code ms-toolsai.jupyter-4 completed in 17ms
10:25:03.863 [debug] Jupyter completion time: 18
10:25:03.974 [debug] Jupyter completion time: 30
10:25:03.978 [debug] Execution of code ms-toolsai.jupyter-5 completed in 33ms
10:25:29.471 [debug] Handle Execution of Cells 0,1,2,3,4,5,6,7,8,9,10 for ~\Documents\Evaluation_Results_2025\Symb_Regression_2025_TSA.ipynb
10:25:29.471 [trace] Execute Notebook ~\Documents\Evaluation_Results_2025\Symb_Regression_2025_TSA.ipynb. Step 1
10:25:29.471 [trace] Connect to Kernel ~\Documents\Evaluation_Results_2025\Symb_Regression_2025_TSA.ipynb. Step 2
10:25:29.472 [trace] Connected to Kernel ~\Documents\Evaluation_Results_2025\Symb_Regression_2025_TSA.ipynb. Step 3
10:25:29.472 [trace] executeCell 0. Step 4
10:25:29.472 [trace] executeCell 1. Step 4
10:25:29.472 [trace] executeCell 2. Step 4
10:25:29.472 [trace] executeCell 3. Step 4
10:25:29.472 [trace] executeCell 4. Step 4
10:25:29.472 [trace] executeCell 5. Step 4
10:25:29.472 [trace] executeCell 6. Step 4
10:25:29.472 [trace] executeCell 7. Step 4
10:25:29.472 [trace] executeCell 8. Step 4
10:25:29.472 [trace] executeCell 9. Step 4
10:25:29.472 [trace] executeCell 10. Step 4
10:25:29.473 [trace] Cell Index:0 sent to kernel
10:25:29.479 [trace] Start cell 0 execution @ 1750256729478 (clear output)
10:25:29.479 [debug] Kernel acknowledged execution of cell 0 @ 1750256729478
10:25:29.699 [trace] Cell 0 completed in 0.221s (start: 1750256729478, end: 1750256729699)
10:25:29.700 [trace] Cell 0 executed successfully
10:25:29.702 [trace] Cell Index:1 sent to kernel
10:25:29.709 [trace] Start cell 1 execution @ 1750256729709 (clear output)
10:25:29.709 [debug] Kernel acknowledged execution of cell 1 @ 1750256729709
10:25:29.717 [trace] Cell 1 completed in 0.008s (start: 1750256729709, end: 1750256729717)
10:25:29.718 [trace] Cell 1 executed successfully
10:25:29.720 [trace] Cell Index:2 sent to kernel
10:25:29.725 [trace] Start cell 2 execution @ 1750256729725 (clear output)
10:25:29.725 [debug] Kernel acknowledged execution of cell 2 @ 1750256729725
10:25:37.163 [trace] Cell 2 completed in 7.438s (start: 1750256729725, end: 1750256737163)
10:25:37.164 [trace] Cell 2 executed successfully
10:25:37.166 [trace] Cell Index:3 sent to kernel
10:25:37.175 [trace] Start cell 3 execution @ 1750256737175 (clear output)
10:25:37.175 [debug] Kernel acknowledged execution of cell 3 @ 1750256737175
10:25:37.252 [trace] Cell 3 completed in 0.076s (start: 1750256737175, end: 1750256737251)
10:25:37.252 [trace] Cell 3 executed successfully
10:25:37.253 [trace] Cell Index:4 sent to kernel
10:25:37.257 [trace] Start cell 4 execution @ 1750256737257 (clear output)
10:25:37.258 [debug] Kernel acknowledged execution of cell 4 @ 1750256737257
10:25:37.264 [trace] Cell 4 completed in 0.007s (start: 1750256737257, end: 1750256737264)
10:25:37.264 [trace] Cell 4 executed successfully
10:25:37.265 [trace] Cell Index:5 sent to kernel
10:25:37.269 [trace] Start cell 5 execution @ 1750256737269 (clear output)
10:25:37.269 [debug] Kernel acknowledged execution of cell 5 @ 1750256737269
10:25:37.280 [trace] Cell 5 completed in 0.011s (start: 1750256737269, end: 1750256737280)
10:25:37.280 [trace] Cell 5 executed successfully
10:25:37.281 [trace] Cell Index:6 sent to kernel
10:25:37.286 [trace] Start cell 6 execution @ 1750256737286 (clear output)
10:25:37.286 [debug] Kernel acknowledged execution of cell 6 @ 1750256737286
10:27:58.734 [debug] creating new code watcher with matching document file:///c%3A/Users/<username>/Documents/TRISO_AI/Coating_Particle_Code/TRISO_venv/Lib/site-packages/pysr/sr.py
10:30:10.787 [debug] creating new code watcher with matching document file:///c%3A/Program%20Files/WindowsApps/PythonSoftwareFoundation.Python.3.11_3.11.2544.0_x64__qbz5n2kfra8p0/Lib/contextlib.py
10:41:57.558 [debug] creating new code watcher with matching document file:///c%3A/Users/<username>/Documents/TRISO_AI/Coating_Particle_Code/TRISO_venv/Lib/site-packages/pysr/sr.py
```

**Reply by MilesCranmer** [reply]
What is the log when you restart the kernel? I want to see where it gets stuck

**Reply by dkandil** [reply]
@MilesCranmer This happens to me too quite often, I am running it on colab, and sometimes it just stop proceeding almost forever and I have to restart the runtime.

**Reply by MilesCranmer** [reply]
Colab lets you start a terminal in the background: can you try this, and see if the outputs folder is still being updated? Or maybe it just completely froze? You could also check the CPU usage when this happens.

Colab specifically might be a different issue. I highly recommend running it locally if you can, as the Colab CPUs tend to be quite weak. A standard laptop will often be faster

**Reply by dkandil** [reply]
Thanks a lot, I just realised how fast the package is outside colab!
